src/lib/permutations/ascon_perm/ascon_perm.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+* Permutation Ascon_p[rounds] as specified in NIST SP.800-232, Section 3
+* (C) 2025 Jack Lloyd
+*     2025 René Meusel
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/ascon_perm.h>
+
+#include <botan/internal/loadstor.h>
+#include <botan/internal/rotate.h>
+#include <botan/internal/stl_util.h>
+
+namespace Botan {
+
+void Ascon_p::absorb(std::span<const uint8_t> input, std::optional<uint8_t> permutation_rounds) {
+   BufferSlicer input_slicer(input);
+
+   // Block-wise incorporation of the input data into the sponge state until
+   // all input bytes are processed
+   while(!input_slicer.empty()) {
```

Comment:
This could be optimized for the case where there is no data currently accumulated and at least a full buffer available. Maybe not worth bothering with at least initially though

Reply:
:+1: Yes fine with me